presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MergeAdjacentWindows.java

```diff
@@ -46,7 +47,16 @@
     @Override
     public Optional<PlanNode> apply(WindowNode parent, Captures captures, Context context)
     {
-        WindowNode child = captures.get(CHILD);
+        // Pulling the descendant WindowNode above projects is done as a part of this rule, as opposed in a
+        // separate rule, because that pullup is not useful on its own, and could be undone by other rules.
+        // For example, a rule could insert a project-off node between adjacent WindowNodes that use different
+        // input symbols.
```

Comment:
> 2. A standalone WindowNode pull-up rule moves WC above P, adding in an identity projection for D.

To me it does not seem to be a good rewrite, notice that P does not produce symbols which are needed by WC. They cannot be swapped. Changing assignments to me is hacky, and I am not yet sure if it is a good approach.

Anyway, If there is a conflict between window merging rules and column pruning rules then I think we need to just run them in separated iterative optimizers for a while.

We strive to support Memo with multiple nodes per group, so such alternatives that P is over WC or WC is over P will become possible in future.